dev/camdenhells/REQUIREMENTS.md

```diff
@@ -55,34 +55,37 @@ For reviews and updates:
 ## 3.1 Android and OpenGL ES integration
 
 1. Provide (major and minor) version query for the current GL ES render context
-2. Allow multiple `GlSurfaceViews` per Activity ? [2.1]
-  * Is this actually worthwhile?
-  * Only one render thread/context/Renderer can exist at once, so SurfaceViews are mutually exclusive 
-    * E.g. `ViewPager` support (per #1619)
-    * Is timing of render context switch still based on window attachment?
-3. Wallpaper surface view multisampling fix (#1559)
-4. Provide query/configuration constants (bit flags) for all nVidia GL ES extensions (#1742) [2.1]
-  * Do any other vendors or extension categories need this?
+2. Decouple Android views from the EGL context to allow offline rendering/initialization
```

Comment:
I'm not at all sure what offline rendering means, and it seems like EGL contexts must be tightly coupled to GlSurfaceView/GlSurfaceTextureView (and so GLThread) implementations.  Is the real goal here just to support off-screen rendering?

Reply:
The nature of what EGL is does not require a true user facing window. Some implementations do and I'm not yet sure if Android is among them. If you use the built in classes of Android then yes, it's required but we already don't use the Android ones, so we could potentially have a shot at this. 

Yes the goal is to support off screen renders. For example, I originally got into rajawali because I was working on a weather app that wished to render an atmospheric shader that calculated Raleigh scattering. This doesn't change much per second or even. Per minute so the goal was to basically render an image every so often, and then use that image in an image view. This never happened but it is this kind of situation I wish to support

Reply:
Cool use case, didn't think of that.  Essentially one-shot on-demand rendering to a bitmap. Should be orthogonal to current GlSurfaceView/GlThread usage, no? Seems like a 2.1 req?

Reply:
I have never been happy with the implementations of rajawali's surface/texture views. From the day I added them to the library I havn't trusted them completely. The surface view uses Android's under the hood and the texture view is basically a clone of what the surface view does. My distrust largely comes from how complex the classes are. I would like to examine the prospects of factoring out the GL thread. There is another reason for this. The GL thread runs in the view and we run a Rajawali thread on top of that currently. Really all this rajawali thread does is tell the GL thread to render a frame, rather than using the built in mechanisms for rendering at a certain frame rate or continuously. This is super wasteful. I am not sure to what extent we can accomplish this but it is definately a major shift in how things operate and therefore a key 2.0 requirement in my mind.

Reply:
I don't think Android's GlSurfaceView/GlThread offers actual frame timing, only continuous vs on-request modes, so I don't know how we could get away from using a separate timer/thread/executor or whatever to generate the render-frame requests. And not sure how we could get around it even if we wrote our own render thread, as it still has to handle surface size change, loss of context, and queued events, but maybe there's a way...

Reply:
Actually, after thinking about it, it should be possible and relatively straightforward to support on-demand off-screen rendering without significant architectural changes such as decoupling from SurfaceViews. If an app is already using a SurfaceView, it should be easy to add a single frame off-screen render whenever needed; if the app is not using a Surface view, it should be easy to just create a hidden instance and use it to do the off-screen render.

Reply:
Correct, but I would like this to be a fallback solution. If at all possible I would like to better control the egl context creation.